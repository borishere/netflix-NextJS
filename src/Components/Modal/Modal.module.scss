.modal {
  position: fixed;
  display: flex;
  justify-content: center;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  box-shadow: 0 0 2px rgba(0, 0, 0, 0.1), 0 2px 10px rgba(0, 0, 0, 0.07), 0 10px 20px rgba(0, 0, 0, 0.05), 0 10px 50px rgba(0, 0, 0, 0.05);
  overflow: hidden;
  z-index: 9999;

  &-content {
    position: relative;
    display: flex;
    flex-direction: column;
    margin: 100px auto 0;
    padding: 60px;
    background: #232323;
    box-shadow: 0 0 2px rgba(0, 0, 0, 0.1), 0 2px 10px rgba(0, 0, 0, 0.07), 0 10px 20px rgba(0, 0, 0, 0.05), 0 10px 50px rgba(0, 0, 0, 0.05);

    .modal-close-btn {
      position: absolute;
      top: 30px;
      right: 30px;
      cursor: pointer;
    }

    .modal-title {
      margin-top: 0;
      font-size: 40px;
      font-weight: 300;
    }

    .modal-body {
      overflow: auto;
    }
  }

  &-buttons {
    margin-top: 60px;
    font-weight: 500;
    line-height: 24px;
    text-align: end;

    .submit-btn {
      background: #f65261;
      border-radius: 4px;
      border: 0;
      padding: 16px;
      margin-left: 13px;
      color: #ffffff;
      width: 180px;
      font-size: 20px;
      cursor: pointer;
    }

    .reset-btn {
      background-color: rgba(0, 0, 0, 0);
      border: 1.5px solid #f65261;
      color: #f65261;
      border-radius: 4px;
      padding: 16px;
      box-sizing: border-box;
      width: 180px;
      font-size: 20px;
      cursor: pointer;
    }
  }
}
