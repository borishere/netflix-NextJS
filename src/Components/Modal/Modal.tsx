import React, { FC } from 'react';
import ReactDOM from 'react-dom';
// @ts-ignore
import closeButton from '../../Images/Close-Button.svg';
import './style.scss';

export interface ModalProps {
  isShown: boolean;
  show: (val: boolean) => void;
  modalClass?: string;
}

export const Modal: FC<ModalProps> = ({ isShown, show, children, modalClass }) => {
  if (!isShown) {
    return null;
  }

  return (
    ReactDOM.createPortal(
      <div className={`modal ${modalClass? modalClass : ''}`}>
        <div className='modal-content'>
          <img
            className='modal-close-btn'
            src={closeButton}
            onClick={() => show(false)}
            aria-label='close'
          />
          <div className='modal-body'>
            {children}
          </div>
        </div>
      </div>,
      document.getElementById('root') as Element
    ));
};