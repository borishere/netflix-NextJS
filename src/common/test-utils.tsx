import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };