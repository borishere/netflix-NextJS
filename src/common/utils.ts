export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';