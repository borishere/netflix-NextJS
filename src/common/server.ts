import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
