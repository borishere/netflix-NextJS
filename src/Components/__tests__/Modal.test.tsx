import React from 'react';
import { screen, customRender, waitFor } from '../../common/test-utils';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { Modal } from '../Modal/Modal';

describe('Modal', () => {
  test('is visible', () => {
    const mockShow = jest.fn();

    const { asFragment } = customRender(
      <Modal
        modalClass='add-movie-modal'
        isShown={true}
        show={mockShow}
      >
        <div>content</div>
      </Modal>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('content')).toBeInTheDocument();
  });

  test('is hidden', () => {
    const mockShow = jest.fn();

    customRender(
      <Modal
        modalClass='add-movie-modal'
        isShown={false}
        show={mockShow}
      >
        <div>content</div>
      </Modal>
    );

    expect(screen.queryByText('content')).not.toBeInTheDocument();
  });

  test('Close on X click', async () => {
    const mockShow = jest.fn();

    customRender(
      <Modal
        modalClass='add-movie-modal'
        isShown={true}
        show={mockShow}
      >
        <div>content</div>
      </Modal>
    );

    userEvent.click(screen.getByLabelText('close'));

    await waitFor(() => {
      expect(mockShow).toHaveBeenCalledWith(false);
    });
  });
});