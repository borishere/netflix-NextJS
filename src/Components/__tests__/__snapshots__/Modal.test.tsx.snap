// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal is visible 1`] = `
<DocumentFragment>
  <div
    class="modal add-movie-modal"
  >
    <div
      class="modal-content"
    >
      <img
        aria-label="close"
        class="modal-close-btn"
        src="test-file-stub"
      />
      <div
        class="modal-body"
      >
        <div>
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
